### ISSUE TITLE
GitHub Action for Service Mesh Patterns

### ISSUE THREAD

**Issue by leecalcote** [service-mesh-patterns/service-mesh-patterns#15]
#### Current Behavior
There are a couple GitHub Actions for Meshery CLI interactions currently, but not a GitHub Action specifically for service mesh patterns.

#### Desired Behavior
 It would be nice to have a GitHub Action dedicated to use cases of importing and applying service mesh patterns.

#### Implementation
Build a new action here - https://github.com/layer5io/mesheryctl-service-mesh-patterns-action

#### Acceptance Tests
1. Users can import service mesh patterns.
1. Users can apply service mesh patterns.

---
**[Contributor Guide](https://docs.meshery.io/project/contributing)**

**Comment by hershd23** [comment]
I am interested to take up this issue.

Note:- I do not have a lot of experience with GitHub actions but would be looking to pick up some knowledge along the way

**Comment by leecalcote** [comment]
@hershd23, thank you for both volunteering and for signposting expectations based on your current level of experience. Let's give this a shot and see how far you get. @theBeginner86, perhaps, you might collaborate with @hershd23? You can copy much of what is written into other `mesheryctl` GitHub Actions.

**Comment by theBeginner86** [comment]
Sure! I'd be happy to collaborate 👍

// @hershd23 @leecalcote

**Comment by hershd23** [comment]
One thing I noticed here is that this action still refers a fork made by a contributor. Should we update this with the layer5io repo?

https://github.com/service-mesh-patterns/service-mesh-patterns/blob/f93ba3b079f2a82a466eadac7abc08a2e29125f8/.github/workflows/pattern-action.yml#L19

**Comment by sayantan1413** [comment]
Hey @hershd23 you are right we need to change the sayantan1413/... to layer5io/....

**Comment by theBeginner86** [comment]
@hershd23 
I'm not sure of the best way of doing it but I generally test out small sections of the GitHub action in one of my testing repo. 😅

Though asking this question in the discussion forum https://discuss.layer5.io, would help us get a better answer from other community members.

**Comment by hershd23** [comment]
Right makes sense, I'll ask this question in the discussion forum

**Comment by hershd23** [comment]
@leecalcote @theBeginner86 @sayantan1413 
A couple of questions :- 

* Is the `import` command available in mesheryctl yet? I am currently testing the existing action for the `apply` command and added the `import` command before it to test it once but it errored out as you can check [here](https://github.com/hershd23/service-mesh-patterns/runs/4908239232?check_suite_focus=true#step:3:1434). I also checked the pattern directory in mesheryctl and it doesn't look like `import` is there yet. Atleast in the master branch.
* Secondly is the idea to create two specific actions for testing `import` and `apply`? Given `import` would a lot of the same setup as that of the `apply` command.

**Comment by leecalcote** [comment]
@piyushsingariya @hexxdump @alphax86

https://docs.meshery.io/reference/mesheryctl/pattern/apply

**Comment by hershd23** [comment]
Any updates? @utkarsh-pro @leecalcote

**Comment by leecalcote** [comment]
@utkarsh-pro , please do chime in when you have a moment.

**Comment by utkarsh-pro** [comment]
@hershd23 right now, both "import" and "apply" happen in the same command. Before we understand how it is done through `mesheryctl` let's clarify what "importing" and "applying" means in the context of patterns.

**Import**: When you _import_ a pattern, you are basically fetching a pattern from a source (here source could be: any HTTP endpoint, filesystem or Github - yes, we have special import features for Github) and saving it on a Meshery Providers database for future retrievals/usage.

**Apply**: Patterns are nothing but YAMLs which are declarative in nature, just like everything is in Kubernetes. When you send a pattern to meshery server for it to process the patterns and perform actions based on the YAML, we call it "applying". It is very similar to `kubectl apply -f`.

Here's how are the ways in which it works:
- A user can run `mesheryctl pattern apply -f <file name | URL>`, this will do the following:
    - `mesheryctl` will "import" the pattern.
    - After "importing", it will "apply" the pattern.
- A user can run `mesheryctl pattern apply <pattern-name>`, this will do the following:
    - `mesheryctl` will search patterns in the provider's database which are registered against the user running the command.
    - If there are more than one pattern registered with the same name then the user is presented with the list to select a pattern.
    - Selected pattern is "applied"
 
It should be noted that a user can pass `--skip-save` in conjunction with above commands, which will skip the "import" step and will only "apply" the pattern.

I hope this clarifies some of the doubts.

**Comment by hershd23** [comment]
Thanks @utkarsh-pro, this makes pattern import and apply quite clear.

This is what I have done till now :- 
https://github.com/hershd23/service-mesh-patterns/runs/4953284735?check_suite_focus=true
Here is a CI run. At this point in time I have changed the `pattern-action.yml` file to the sample action format given [here](https://github.com/layer5io/meshery-service-mesh-patterns-action#deploying-a-pattern-with-url) with the correct action location. This runs successfully, and at this point runs for any URL supplied.

Logically the next question is what would the deliverables be for this issue. @leecalcote @theBeginner86 do help me out here
1. The part of testing out various service mesh patterns could be done easily, with this workflow asking for various parameters before the run. Simply providing the URL of the pattern would suffice here.
2. How should we go about implementing the second case as described by @utkarsh-pro? One (not so neat) way to do this is to apply pattern1 using URL then pattern2 using URL and then trying to apply pattern1 again using only the `pattern-name`. Let me know if there's a better idea

**Comment by leecalcote** [comment]
@utkarsh-pro 💪 

@hershd23, chat about this in today's Meshery Build and Release meeting? Details here - https://meet.layer5.io

**Comment by leecalcote** [comment]
An example use - https://github.com/meshery/meshery/blob/620ff1ef940af18cbfd846e5fe20c68bffb7b169/.github/workflows/testadapters.yaml#L143

**Comment by hershd23** [comment]
A couple of comments on this issue. I went through the CI call and understood there was some uncertainty whether this is required or not.

A lot of what we want is already done here https://github.com/layer5io/meshery-service-mesh-patterns-action/blob/master/mesheryctl.sh. The structure of the action is a little convoluted though. 

If you see the repo [here](https://github.com/layer5io/meshery-service-mesh-patterns-action/blob/master), this edited[workflow script](https://github.com/hershd23/service-mesh-patterns/blob/pattern-import-apply-ci/.github/workflows/pattern-action.yml) calls [action.yml](https://github.com/layer5io/meshery-service-mesh-patterns-action/blob/master/action.yml) which in turn calls [main.js](https://github.com/layer5io/meshery-service-mesh-patterns-action/blob/master/main.js) which in turn spins up a child process [main.sh](https://github.com/layer5io/meshery-service-mesh-patterns-action/blob/master/main.sh) which then calls [meshery.sh](https://github.com/layer5io/meshery-service-mesh-patterns-action/blob/master/meshery.sh) for setup actions like setting up minikube and meshery and then [mesheryctl.sh](https://github.com/layer5io/meshery-service-mesh-patterns-action/blob/master/mesheryctl.sh) which tests the 'mesheryctl pattern apply -f <file-name>' action.

TL;DR :- Most of what we need is already there in https://github.com/layer5io/mesheryctl-service-mesh-patterns-action, just in a way that is a little convoluted. 

Hence just changing the `pattern-action.yml` workflow to the format given [here](https://github.com/layer5io/meshery-service-mesh-patterns-action#deploying-a-pattern-with-url), the whole action runs successfully. **Even if this not a must have** we can easily just change [`pattern-action.yml`](https://github.com/service-mesh-patterns/service-mesh-patterns/blob/master/.github/workflows/pattern-action.yml) and we'll have a working action for pattern apply

**Comment by leecalcote** [comment]
@hershd23 great research. Please proceed. As reusable Action, the Patterns Action shouldn't provision Kubernetes, but ask the user for their kubeconfig. This would be a good change to make.
